design_notes/XDN05 - xlang Binary Interface.md

```diff
@@ -12,6 +12,233 @@ status: draft
+the correct COM interface pointer type.
+
+Methods on COM interfaces (including xlang ABI interfaces) never throw C++ exceptions, due to the
+need for a stable contract across compilers. QueryInterface returns a 32-bit signed integer as an
+error code known. On Windows, this code is known as an
+[HRESULT](https://docs.microsoft.com/en-us/windows/desktop/com/error-handling-in-com).
+Generally with HRESULTs, negative integers indicate errors, positive integers and zero indicate
+success. QueryInterface in particular should always return 0 on success.
+
+QueryInterface may return the following error codes:
+
+- 0x80004002 (E_NOINTERFACE) indicates the requested interface is not supported by this object.
+
+### AddRef and Release
+
+COM objects manage their own lifetime via an internal reference count. When this count goes to
+zero, the COM object cleans itself up. IUnknown provides methods to increment and decrement the
+object's reference count. If a language projection makes a copy of a COM interface pointer, it must
+call AddRef. When client code is finished using a given pointer to a COM interface, the language
+projection must call Release.
+
+The return value of these function is specified to be the new reference count of the object.
+However the COM standard itself warns that this value is intended to be used only for test purposes.
+In practice on Windows, the value returned from these two functions is often inaccurate. The one
+scenario where you can trust the return value is when Release returns zero - in this case, you can
+reliably assume the object has cleaned itself up.
+
+## IXlangObject
+
+In addition to the core functionality described by IUnknown, ABI interfaces for xlang types derive
+from IXlangObject. This interface adds additional core functionality needed for xlang.
+
+``` cpp
+enum ObjectInfoCategory
+{
+    StringRepresentation,
+    TypeName,
+    MemoryUsage
+};
+
+struct IXlangObject
+{
+    int32_t GetInfo(ObjectInfoCategory info_category, void** info);
+};
+```
+
+### GetInfo
+
+In WinRT, the IInspectable interface provides property get methods that retrieved information about
+the object needed in specific runtime scenarios. Over time, the list of information categories that
+might be useful at runtime grew. For example, knowing the memory footprint of a given WinRT object
+would allow for garbage collected languages like C# make better scheduling decisions. But since
+IInspectable is the base COM interface for every interface in WinRT, it's impossible to extend.
+
+For xlang, instead of individual property get methods, IXlangObject exposes a single GetObjectInfo
+method. The method takes an enum value indicating the object info being requested and a void** to
+hold the returned value. xlang objects are not required to support all object information
+categories. Similar to QueryInterface, an object can simply return an error code indicating that a
```

Comment:
specify that error?  e.g., E_NOTIMPL

Reply:
I don't think we want a request for an unsupported category to cause an exception.  That leads to a horrible developer experience, especially as we're saying objects don't need to support all categories.

Thus we should never return a failure result.

Reply:
Brent, this function is never projected. A language projection calls it internally and thus there's no concern about an exception. Of course, I'm happy with a simpler "bool GetObjectInfo(category, void**) signature as well. An error code is overrated here.

Reply:
I like making it a boolean